primefaces/src/main/java/org/primefaces/util/ComponentUtils.java

```diff
@@ -632,7 +634,24 @@ public static void encodeIndexedId(FacesContext context, UIComponent component,
         // append index to all id's
         char separator = UINamingContainer.getSeparatorChar(context);
         String encodedComponent = fsw.toString();
-        encodedComponent = encodedComponent.replaceAll("\\sid=\"(.*?)\"", " id=\"$1" + separator + index + "\"");
+
+        // find all id's to replace
+        Pattern pattern = Pattern.compile("\\sid=\"(.*?)\"");
+        Matcher matcher = pattern.matcher(encodedComponent);
+
+        // grab all the unique ID's found
+        Set<String> ids = new HashSet<>(5);
+        while (matcher.find()) {
+
+            ids.add(matcher.group(1));
+        }
+
+        // replace each id with an indexed version
+        for (String id : ids) {
```

Comment:
Shouldn't this be an index loop? https://github.com/primefaces/primefaces/blob/master/CONTRIBUTING.md#performance-guidelines

Reply:
True but its not an ArrayList its a HashSet for uniqueness and there is no way to iterator over a HashSet without an iterator?